linux/preempt.c

```diff
@@ -15,14 +15,25 @@
  * correct to instead guarantee mutual exclusion for the critical sections.
  */
 
-static pthread_mutex_t preempt_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
+static pthread_mutex_t preempt_lock;
+static pthread_once_t once_control = PTHREAD_ONCE_INIT;
+
+void init_mutex(void) {
+	pthread_mutexattr_t attr;
+	pthread_mutexattr_init(&attr);
+	pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+	pthread_mutex_init(&preempt_lock, &attr);
+	pthread_mutexattr_destroy(&attr);
+}
 
 void preempt_disable(void)
 {
+	pthread_once(&once_control, init_mutex);
```

Comment:
This is a hot path, let's not do it here.

just mark init_mutex with __attribute__((constructor)), we're already using that

Reply:
done. not sure what priority to use here, but i don't see other constructors calling preemt_* functions, so i left it blank.